slackbot_using_azure_openai/service.bal

```diff
@@ -0,0 +1,88 @@
+import ballerina/crypto;
+import ballerina/http;
+import ballerinax/azure.openai.chat;
+
+configurable string openAIToken = ?;
+configurable string serviceUrl = ?;
+configurable string deploymentId = ?;
+configurable string slackSigningSecret = ?;
+
+const API_VERSION = "2023-03-15-preview";
+const MAX_MESSAGES = 25;
+
+final chat:Client azureOpenAI = check new (
+    config = {auth: {apiKey: openAIToken}},
+    serviceUrl = serviceUrl
+);
+
+type ChatMessage record {|
+    string role;
+    string content;
+|};
+
+type Response record {|
+    string response_type;
+    string text;
+|};
+
+enum ROLE {
+    SYSTEM = "system",
+    USER = "user",
+    ASSISTANT = "assistant"
+}
+
+service /slack on new http:Listener(8080) {
+    map<ChatMessage[]> chatHistory = {};
+
+    resource function post events(http:Request request) returns Response|error {
+        if !check verifyRequest(request) {
+            return error("Request verification failed");
+        }
+
+        map<string> requestPayload = check request.getFormParams();
+
+        string? channelName = requestPayload["channel_name"];
+        string? requestText = requestPayload["text"];
+        if channelName is () || requestText is () {
+            return error("Invalid values in the request payload for channel_name or text");
+        }
+
+        ChatMessage[] history = self.chatHistory[channelName] ?:
+                                [{role: SYSTEM, content: "You are an AI slack bot to assist with user questions."}];
+        history.push({role: USER, content: requestText});
+
+        chat:Chat_completions_body chatBody = {messages: history};
+        chat:Inline_response_200 completion = check azureOpenAI->/deployments/[deploymentId]/chat/completions.post(
+            API_VERSION, chatBody
+        );
+
+        chat:Inline_response_200_message? response = completion.choices[0].message;
+        if response is () {
+            return error("Error in response generation");
+        }
+
+        // Limit history to 25 messages to preserve token limit.
+        if history.length() > MAX_MESSAGES {
+            history = history.slice(1, history.length());
```

Comment:
```suggestion
            history = history.slice(0, MAX_MESSAGES);
```

If the intention is to limit to 25 messages isn't the above code correct?

Reply:
We want to remove the earliest message first. In the suggested code, I believe we'd be truncating from the tail end of the array. Isn't that so?

Reply:
@JayaniH If we want to remove the fist N elements from the history to limit to `MAX_MESSAGES`, shouldn't we do as 
`history = history.slice(history.length() - MAX_MESSAGES);`? 

here we don't need to provide the `endIndex` argument as its default value will be the array length anyway